homeassistant/helpers/entity.py

```diff
@@ -1223,23 +1220,26 @@ def _async_write_ha_state(self) -> None:
             self._context = None
             self._context_set = None
 
-        try:
-            hass.states.async_set_internal(
-                entity_id,
+        if len(state) > MAX_LENGTH_STATE_STATE:
+            _LOGGER.error(
+                "State %s for %s is longer than %s, falling back to %s",
                 state,
-                attr,
-                self.force_update,
-                self._context,
-                self._state_info,
-                time_now,
-            )
-        except InvalidStateError:
-            _LOGGER.exception(
-                "Failed to set state for %s, fall back to %s", entity_id, STATE_UNKNOWN
-            )
-            hass.states.async_set(
```

Comment:
AFAICT, It didn't make so much sense to fallback to `async_set` here as the only reason we would raise `InvalidStateError` is the state was too long